// .ignored/notchpay-client/dist/usecases/payment.js
import { NotchPayChannel, } from "../types";
import { Usecase } from "./base";
export class PaymentUsecases extends Usecase {
    /**
     * @returns Promise with all payments you need
     */
    async getAll() {
        const response = await this.client.get("/payments");
        return response.data;
    }
    /**
     * Initialize a payment (legacy method)
     * @param paymentData Payment initialization data
     * @returns Promise with payment response
     */
    async initialize(paymentData) {
        const response = await this.client.post("/payments/initialize", paymentData);
        return response.data;
    }
    /**
     * Create a payment (modern method)
     * @param paymentData Payment data
     * @returns Promise with payment response
     */
    async create(paymentData) {
        const response = await this.client.post("/payments", paymentData);
        return response.data;
    }
    /**
     * Initiate a direct charge for an existing transaction
     * @param ref The transaction reference returned from payment initialization
     * @param chargeData Direct charge data including channel and payment method details
     * @returns Promise with direct charge response
     */
    async directCharge(ref, chargeData) {
        const response = await this.client.put(`/payments/${ref}`, chargeData);
        return response.data;
    }
    /**
     * Get payment info
     * @param reference The transaction reference
     * @returns Promise with transaction details
     */
    async get(reference) {
        const response = await this.client.get(`/payments/${reference}`);
        return response.data;
    }
    /**
     * Initialize Mobile Money payment in one step (convenience method)
     * @param paymentData Initial payment data
     * @param phoneNumber Mobile Money phone number
     * @param channel Payment channel (MTN, Orange, etc)
     * @returns Promise with complete payment flow result
     */
    async initializeMobileMoneyPayment(paymentData, phoneNumber, channel = NotchPayChannel.MOBILE) {
        // First initialize the payment
        const initResponse = await this.create(paymentData);
        // Extract transaction reference from the response
        const transactionId = initResponse.transaction.reference;
        // Proceed with direct charge using mobile money
        return this.directCharge(transactionId, {
            channel: channel,
            data: {
                phone: phoneNumber,
            },
        });
    }
}
